src/arch/arm/kernel/boot.c

```diff
@@ -326,25 +325,6 @@ static BOOT_CODE bool_t try_init_kernel(
     bi_frame_vptr = ipcbuf_vptr + BIT(PAGE_BITS);
     extra_bi_frame_vptr = bi_frame_vptr + BIT(PAGE_BITS);
 
-    /* If no DTB was provided, skip allocating extra bootinfo */
-    p_region_t dtb_p_reg = {
-        dtb_addr_start, ROUND_UP(dtb_addr_end, PAGE_BITS)
```

Comment:
Rounding up the dtb_addr_end to page alignment looks to be unnecessary here.  This region of memory is only tracked by the kernel in order to give it to user level in an extra boot info field.  The extra boot info datastructure is a list of variable length blocks, with a seL4_BootInfoHeader at the start followed by a variable length of data. No guarantees are made about either the start of the bootinfo header, or the start of the data that comes after it being page aligned.  If alignment is required, then there is an explicit SEL4_BOOTINFO_HEADER_PADDING type for adding an entry that is used to add padding so that the next entry can have a certain alignment.  Rounding the dtb size up to a PAGE almost guarantees that the bootinfo extra size is going to use an additional page of memory because of `extra_bi_size = sizeof(seL4_BootInfoHeader) + (dtb_reg.end - dtb_reg.start);` further along.

Reply:
Also it's odd that the memory for the DTB is marked as reserved when new memory is allocated for the DTB to be copied into and the original memory is never released. The original DTB region should ideally be free'd and  given back to userlevel as untypeds like how the boot code and data segments are.


(Both of these things are thoughts for future changes as they are extra to any clean up you are trying to do here.)

Reply:
Maybe reserving the area is intended for the situation where a DTB is passed from a previous bootloader transparently by the ELF Loder (instead of using the DTB from the CPIO archive). Since the ELF loader does not tell us if it provided the DTB or just passes one on, reserving this area seems not a bad idea. We an add a comment here with this explanation. In the long adding some adding some flags to the boot interface about the DTB origin might be an option.

Reply:
Or maybe this it just intended to protect the DTB from getting overwritten accidentally by the kerne memory initialization when it needs to allocate memory for internal purposes. Since the DTB from the ELF Loader is likely in an area that is considered free memory, this might happen. What we could do here is recycle the memory like we already do with the kernel boot memory,